src/nxt_fs.c

```diff
@@ -273,6 +273,32 @@ nxt_fs_mkdir_all(const u_char *dir, mode_t mode)
 }
 
 
+nxt_int_t
+nxt_fs_mkdir_dirname(const u_char *path, mode_t mode)
+{
+    int   err;
+    char  *ptr;
+    char  *dir;
+
+    dir = nxt_strdup(path);
+    if (nxt_slow_path(dir == NULL)) {
+        return NXT_ERROR;
```

Comment:
Small issue; you forgot to catch errors.  It is caught here, but ignored later outside of `nxt_fs_mkdir_dirname()`.

Reply:
Or was it on purpose, just because a failure here doesn't really matter?

Reply:
Yes, catch the memory error rather than segfault. But if we don't actually create the directory then that is no different to previously.

Reply:
So in src/nxt_runtime.c::nxt_runtime_exit() there is

```C
    if (rt->type == NXT_PROCESS_MAIN) {                                         
        if (rt->pid_file != NULL) {                                             
            nxt_file_delete(rt->pid_file);                                      
        }                                                                       
                                                                                
#if (NXT_HAVE_UNIX_DOMAIN)                                                      
        {                                                                       
            nxt_sockaddr_t   *sa;                                               
            nxt_file_name_t  *name;                                             
                                                                                
            sa = rt->controller_listen;                                         
                                                                                
            if (sa->u.sockaddr.sa_family == AF_UNIX) {                          
                name = (nxt_file_name_t *) sa->u.sockaddr_un.sun_path;          
                (void) nxt_file_delete(name);                                   
            }                                                                   
        }                                                                       
#endif                                                                          
    }
```

This removes the pid file and control socket.

I could simply create an analogous function to nxt_fs_mkdir_dirname, e.g nxt_fs_rmdir_dirname, that would delete the directories (yes, they will often be the same)...

**However**, I am now convinced that we should **not** do that.

In the cgroups patch I do try and clean up cgroup directories

1) These can change from config to config.
2) I know at which point I should stop removing directories, e.g '/sys/fs/cgroup'

In the control socket/pid file case

1) These should not be changing.
2) I don't really know at which point in the file system hierarchy I should stop removing directories.

For (2) in the case of `/run/unit/control.sock` that could result in three rmdir(2)'s
```
rmdir("/run/unit");
rmdir("/run");
rmdir("/");
```
Of course no real harm will likely be done in this case as '/run' and '/' are most likely not going to be empty, but it just seems a little wrong to try.

On a side note, it's a pity that `rt->listen_sockets->elts` seems to be empty
```gdb
(gdb) p *(nxt_listen_socket_t *)rt->listen_sockets->elts
$3 = {
  socket = 0,
  backlog = 0,
  work_queue = 0x0,
  handler = 0x0,
  sockaddr = 0x0,
  count = 0,
  flags = 0 '\000',
  read_after_accept = 0 '\000',
  ipv6only = 0 '\000',
  socklen = 0 '\000',
  address_length = 0 '\000'
```

But I guess these are set in some other process after being forked from the main process...

Reply:
I agree that we should not even try, even if we can.  It's not possible to know exactly where to stop, and I don't feel happy trying to *rmdir(2)* everything up to `/`.

Also, GNU standards for Makefiles say that a Makefile doesn't need to try to delete directories, because they could have existed previously:

>  There is no need to delete parent directories that were created with ‘mkdir -p’, since they could have existed anyway.
<https://www.gnu.org/prep/standards/html_node/Standard-Targets.html>

That is said for targets that remove stuff, such as `clean`.  All the time, it talks about removing files, which is what makes sense removing.

Also, as we've seen, other programs don't remove dirs either.